src/NServiceBus.Core/MessageInterfaces/MessageMapper/Reflection/MessageMapper.cs

```diff
@@ -250,6 +259,8 @@ static string GetTypeName(Type t)
 
         readonly object messageInitializationLock = new object();
 
+        HashSet<Type> initializedTypes = new HashSet<Type>();
```

Comment:
Given that this is potentially called concurrently from `DeserializeLogicalMessagesConnector` for each incoming message , I don't think we can use just a `HashSet<Type>` here.

I was thinking about just switching to `ConcurrentDictionary<Type, byte>` and just ignore the value, but I also see that we've already got `nameToType` which appears to be doing something similar. Perhaps we could just reuse that in some way instead? That would mean you'd have to call `GetTypeName` every time though.